src/test/sv2_template_provider_tests.cpp

```diff
@@ -0,0 +1,298 @@
+#include <addresstype.h>
+#include <boost/test/unit_test.hpp>
+#include <common/sv2_messages.h>
+#include <node/transaction.h>
+#include <node/sv2_template_provider.h>
+#include <test/util/net.h>
+#include <test/util/setup_common.h>
+#include <test/util/transaction_utils.h>
+#include <util/sock.h>
+#include <util/strencodings.h>
+
+BOOST_FIXTURE_TEST_SUITE(sv2_template_provider_tests, TestChain100Setup)
+
+/**
+  * A class for testing the Template Provider. Each TPTester encapsulates a
+  * Sv2TemplateProvider (the one being tested) as well as a Sv2Cipher
+  * to act as the other side.
+  */
+class TPTester {
+private:
+    std::unique_ptr<Sv2Transport> m_peer_transport; //!< Transport for peer
+    std::unique_ptr<Sock> m_peer_socket;
+
+public:
+    std::unique_ptr<Sv2TemplateProvider> m_tp; //!< Sv2TemplateProvider being tested
+
+    TPTester(ChainstateManager& chainman, CTxMemPool& mempool)
+    {
+        m_tp = std::make_unique<Sv2TemplateProvider>(chainman, mempool);
+    }
+
+    bool start()
+    {
+        bool started = m_tp->Start(Sv2TemplateProviderOptions { .port = 18447 });
+        if (! started) return false;
+        // Avoid "Connection refused" on CI:
```

Comment:
The template provider tests are quite brittle because they use a real socket.

Reply:
For the time being I just added handling for `MSG_MORE` (on e.g. macOS sequential messages are sent separately while on Linux they're combined). I also made the timeouts a bit longer.

Hopefully that does the trick. This can be revisited closer to the time when the Template Provider is ready for its own PR.

Reply:
I should probably look into using `StaticContentsSock`

Reply:
@vasild any thoughts on how to make mock `Sock`s that can be used to play messages in two directions?